homeassistant/helpers/device_registry.py

```diff
@@ -283,44 +346,47 @@ def _async_update_device(
     @callback
     def async_remove_device(self, device_id: str) -> None:
         """Remove a device from the device registry."""
-        del self.devices[device_id]
+        device = self.devices.pop(device_id)
+        self.deleted_devices[device_id] = device
         self.hass.bus.async_fire(
             EVENT_DEVICE_REGISTRY_UPDATED, {"action": "remove", "device_id": device_id}
         )
         self.async_schedule_save()
 
+    @callback
+    def async_purge_deleted_device(self, device_id: str) -> None:
+        """Permanently remove a device from the device registry."""
+        del self.deleted_devices[device_id]
+        self.async_schedule_save()
+
+    @callback
+    def async_restore_device(self, device_id: str) -> None:
+        """Restore a deleted device."""
+        device = self.deleted_devices.pop(device_id)
+        self.devices[device_id] = device
+        self.hass.bus.async_fire(
+            EVENT_DEVICE_REGISTRY_UPDATED, {"action": "create", "device_id": device_id}
+        )
+        self.async_schedule_save()
+
     async def async_load(self):
         """Load the device registry."""
         async_setup_cleanup(self.hass, self)
 
         data = await self._store.async_load()
 
         devices = OrderedDict()
+        deleted_devices = OrderedDict()
 
         if data is not None:
             for device in data["devices"]:
-                devices[device["id"]] = DeviceEntry(
-                    config_entries=set(device["config_entries"]),
-                    connections={tuple(conn) for conn in device["connections"]},
-                    identifiers={tuple(iden) for iden in device["identifiers"]},
-                    manufacturer=device["manufacturer"],
-                    model=device["model"],
-                    name=device["name"],
-                    sw_version=device["sw_version"],
-                    # Introduced in 0.110
-                    entry_type=device.get("entry_type"),
-                    id=device["id"],
-                    # Introduced in 0.79
-                    # renamed in 0.95
-                    via_device_id=(
-                        device.get("via_device_id") or device.get("hub_device_id")
-                    ),
-                    # Introduced in 0.87
-                    area_id=device.get("area_id"),
-                    name_by_user=device.get("name_by_user"),
-                )
+                devices[device["id"]] = deserialize_device_entry(device)
+            if "deleted_devices" in data:
```

Comment:
This is intended for backwards compatibility when loading from store.
Should it be handled as a new version of the store format + migration instead?

Reply:
Alternative is to do `for device in data.get("deleted_devices", []):`